src/compiler/checker.ts

```diff
@@ -22630,7 +22633,153 @@ namespace ts {
+                    // passing all tests and issue #39110/#39114, but does it deserve? Is the code clear enough?
+
+                    // if(isThisTypeParameter(computedType)){
+                    //     return false;
+                    // }
+                    // if(isElementAccessExpression(expr)){
+                    //     // the only purpose of these code is to omit above condition 2.
+                    //     const type = getReducedApparentType(computedType);
+                    //     if (type.flags & TypeFlags.Object) {
+                    //         const resolved = resolveStructuredTypeMembers(<ObjectType>type);
+                    //         if(resolved.stringIndexInfo || resolved.numberIndexInfo){
+                    //             return false;
+                    //         }
+                    //     }
+                    // }
+                    return true;
+                }
+
+                // here are all possiable situations:
+                // 1. nullable | nonNullable1
+                // 2. nullable | nonNullable1 | nonNullable2...
+                // 3. nonNullable1 | nonNullable2...
+                // if root contains nullable type, for situation 1, we do not need to narrow.
+                const computedTypeIsNullable = maybeTypeOfKind(computedType, TypeFlags.Nullable);
+                return isTypeArrayDiscriminant(reachableFinalType.concat(rootNullable), computedTypeIsNullable);
+
+                // main part is copied from function createUnionOrIntersectionProperty
+                function isTypeArrayDiscriminant(types: Type[], nullableFlag: boolean) {
+                    let checkFlags = 0;
+                    let firstType: Type | undefined;
+                    for (const type of types) {
+                        if (nullableFlag && (type.flags & TypeFlags.Nullable)) {
+                            continue;
+                        }
+                        if (!firstType) {
+                            firstType = type;
+                        }
+                        else if (type !== firstType) {
+                            checkFlags |= CheckFlags.HasNonUniformType;
+                        }
+                        if (isLiteralType(type)) {
+                            checkFlags |= CheckFlags.HasLiteralType;
+                        }
+                        if (type.flags & TypeFlags.Never) {
+                            checkFlags |= CheckFlags.HasNeverType;
+                        }
+                    }
+                    // next line is copied from isDiscriminantProperty
+                    return (checkFlags & CheckFlags.Discriminant) === CheckFlags.Discriminant && !maybeTypeOfKind(getUnionType(types), TypeFlags.Instantiable);
+                }
+            }
+
+            function narrowTypeByDiscriminantNew(type: Type, access: AccessExpression, narrowTypeCb: (t: Type) => Type): Type {

```

Comment:
Why isn’t this (and its `isMatchingReferenceDiscriminant` counterpart) used in `narrowTypeByTypePredicate`? That’s the only remaining use of the old version. Obviously we’re not going to have one function called `narrowTypeByDiscriminant` and one called `narrowTypeByDiscriminantNew`, so what’s the plan here?

Reply:
Whops, You are absolutely right!
The origional purpose is to save the cost. The new function is a little more expensive than the old one. But when time goes on, more and more old function have to be placed with the new one. There is no need to keep the only old version more.